orderbook/src/api/post_quote.rs

```diff
@@ -79,33 +179,261 @@ struct OrderQuote {
     buy_token_balance: BuyTokenDestination,
 }
 
+#[derive(Debug)]
+pub enum FeeAndQuoteError {
+    Fee(FeeError),
+    Order(ValidationError),
+    // RateLimit, // TODO - use this.
+    Other(anyhow::Error),
+}
+
+impl FeeAndQuoteError {
+    pub fn convert_to_reply(self) -> (Json, StatusCode) {
+        match self {
+            FeeAndQuoteError::Fee(err) => err.to_warp_reply(),
+            FeeAndQuoteError::Order(err) => err.to_warp_reply(),
+            // FeeAndQuoteError::RateLimit => (
+            //     super::error("RateLimit", "Too many order quotes"),
+            //     StatusCode::TOO_MANY_REQUESTS,
+            // ),
+            FeeAndQuoteError::Other(err) => {
+                tracing::error!(?err, "get_fee_and_quote error");
+                (super::internal_error(), StatusCode::INTERNAL_SERVER_ERROR)
+            }
+        }
+    }
+}
+
+impl OrderQuoteRequest {
+    /// This method is used by the old, deprecated, fee endpoint to convert {Buy, Sell}Requests
+    pub fn new(sell_token: H160, buy_token: H160, side: OrderQuoteSide) -> Self {
+        Self {
+            sell_token,
+            buy_token,
+            side,
+            ..Default::default()
+        }
+    }
+
+    fn build_quote_from_fee_response(
+        &self,
+        fee_response: FeeResponse,
+    ) -> Result<OrderQuote, FeeAndQuoteError> {
+        let values_result = match (self.side, fee_response.side) {
+            (
+                OrderQuoteSide::Sell { sell_amount },
+                OrderQuoteSide::Buy {
+                    buy_amount_after_fee,
+                },
+            ) => {
+                let sell_amount = match sell_amount {
+                    SellAmount::BeforeFee {
+                        value: sell_amount_before_fee,
+                    } => sell_amount_before_fee,
+                    SellAmount::AfterFee {
+                        value: sell_amount_after_fee,
+                    } => sell_amount_after_fee
+                        .checked_sub(fee_response.fee.amount)
+                        .ok_or(FeeAndQuoteError::Fee(FeeError::SellAmountDoesNotCoverFee))?
+                        .max(U256::one()),
+                };
+                Ok((sell_amount, buy_amount_after_fee, OrderKind::Sell))
+            }
+            (
+                OrderQuoteSide::Buy {
+                    buy_amount_after_fee,
+                },
+                OrderQuoteSide::Sell { sell_amount },
+            ) => match sell_amount {
+                SellAmount::BeforeFee { value: sell_amount } => {
+                    Ok((sell_amount, buy_amount_after_fee, OrderKind::Buy))
+                }
+                SellAmount::AfterFee { .. } => Err(FeeAndQuoteError::Other(anyhow!(
+                    "Currently unsupported route"
+                ))),
+            },
+            _ => Err(FeeAndQuoteError::Other(anyhow!(
+                "Can't build Quote from (QuoteRequest,FeeResponse) when QuoteSide have same type."
+            ))),
```

Comment:
nit: I'm not a huge fan of having these panics in unreachable routes when the code can be refactored to avoid this. I think it may be because the new route uses the old `calculate_fee` logic and then builds a `QuoteResponse` from that. I would suggest instead building a `QuoteResponse` directly and then converting it to either a `BuyResponse` or `SellResponse`.

Reply:
I believe we just removed Buy Response and Sell Response, but I can reintroduce them... I don't see how converting them will circumvent this faulty pairing. Of course, we would never pass the wrong info, but the fact remains that the QuoteRequest can be of two types that needs to match with the response type.

Reply:
My point is that in `post_quote` module we are first mapping
```
fn(OrderQuoteRequest) -> FeeResponse
```

And then
```
fn(OrderQuoteRequest, FeeResponse) -> OrderQuote
```

This is the cause of the unreachable branch here, because the inputs `OrderQuoteRequest` and `FeeResponse` have to "match". In order to not have this kind of unreachable branch, you could instead of directly implement:
```
fn(OrderQuoteRequest) -> OrderQuote
```

Then, in the `fee_and_quote` module, you need:
```
fn(SellRequest) -> SellResponse
```

And this can be achieved by first implementing:
```
fn(SellRequest) -> OrderQuoteRequest
```

Then calling the common logic implemented in the `post_quote` module for:
```
fn(OrderQuoteRequest) -> OrderQuote
```

And finally, mapping the resulting `OrderQuote` with:
```
fn(OrderQuote) -> SellResponse
```

Reply:
Ok... I will try this out although I am not sure where in the mix the `FeeResponse` plays a role. It seems as if you are suggesting to build the `OrderQuote` without having a `FeeResponse`, but this doesn't seems reasonable and we will also wind with and intermediary between `OrderQuoteRequest` and `OrderQuote` (the FeeResponse) Whose side type will have to be matched with Request's side type...

Reply:
For example, in pseudo-code:
```
    fn calculate_quote(
        &self,
        fee_calculator: Arc<dyn MinFeeCalculating>,
        price_estimator: Arc<dyn PriceEstimating>,
    ) -> Result<OrderQuote, OrderQuoteError> {
        let (sell_amount, buy_amount, fee_amount, kind) = match self.side {
            OrderQuoteSide::Sell { sell_amount: SellAmount::BeforeFee { .. } } => { /* same as current calculate_fee */ }
            OrderQuoteSide::Sell { sell_amount: SellAmount::AfterFee { .. } } => { /* same as current calculate_fee */ }
            OrderQuoteSide::Buy { .. } => { /* same as current calculate_fee */ }
        };
        Ok(OrderQuote {
            from: self.from,
            sell_token: self.sell_token,
            buy_token: self.buy_token,
            receiver: self.receiver,
            sell_amount,
            buy_amount,
            valid_to: self.valid_to,
            app_data: self.app_data,
            fee_amount,
            kind,
            partially_fillable: self.partially_fillable,
            sell_token_balance: self.sell_token_balance,
            buy_token_balance: self.buy_token_balance,
        })
    }
```

This way, you only have a single match on the request `side`, and produce an `OrderQuote`.

You can then in `fee_and_quote` module:
```
impl From<SellQuery> for OrderQuoteQuery { ... }
impl From<OrderQuote> for FeeResult { ... }
```

And do:
```
OrderQuoteRequest::from(sell_query).calculate_quote(...).into()
```

The main argument for doing it this way, is that the current `calcualte_fee` method is very much "in the shape" of the old API (the `fee_and_quote` API). This would make future changes/improvements to the `post_quote` API harder, since you would have to keep some compatibility with the legacy `fee_and_quote` API (instead of, as I suggested, implementing the **legacy** `fee_and_quote` API by "wrapping" the new `post_quote` API).